codalab/lib/bundle_cli.py

```diff
@@ -801,61 +814,71 @@ def do_upload_command(self, args):
-                raise UsageError("Cannot upload from path %s: no filesystem available." % path)
-            else:
-                # Check that the upload path exists.
-                path_util.check_isvalid(path_util.normalize(path), 'upload')
+        if args.contents is not None and args.path:
+            raise UsageError("Upload does not support mixing content strings and local files")
 
-        # If contents of file are specified on the command-line, then include that with the bundles to upload.
-        if args.contents is not None:
-            if not args.md_name:
-                args.md_name = 'contents'
-            tmp_path = tempfile.mkstemp()[1]
-            f = open(tmp_path, 'w')
-            print >>f, args.contents
-            f.close()
-            args.path.append(tmp_path)
-
-        # Canonicalize (e.g., removing trailing /)
-        sources = [path_util.normalize(path) for path in args.path]
-
-        # Pull out the upload bundle type from the arguments and validate it.
-        # Note: only allow dataset bundles (eventually deprecate the program bundle and just have uploaded bundles).
-        bundle_type = 'dataset'
-        bundle_subclass = get_bundle_subclass(bundle_type)
-        metadata = self.get_missing_metadata(bundle_subclass, args, initial_metadata={})
+        # Build bundle info
+        metadata = self.get_missing_metadata(UploadedBundle, args, initial_metadata={})
         # name = 'test.zip' => name = 'test'
+        if args.contents is not None:
+            metadata['name'] = 'contents'
         if not args.pack and zip_util.path_is_archive(metadata['name']):
             metadata['name'] = zip_util.strip_archive_ext(metadata['name'])
+        bundle_info = {
+            'bundle_type': 'dataset',  # TODO: deprecate Dataset and ProgramBundles
+            'metadata': metadata,
+        }
 
-        # Type-check the bundle metadata BEFORE uploading the bundle data.
-        # This optimization will avoid file copies on failed bundle creations.
-        # pass in a null owner to validate. Will be set to the correct owner in the client upload_bundle below.
-        bundle_subclass.construct(owner_id=0, metadata=metadata).validate()
-        info = {'bundle_type': bundle_type, 'metadata': metadata}
-
-        # Finally, once everything has been checked, then call the client to upload.
-        uuid = client.upload_bundle(
-            sources=sources,
-            follow_symlinks=args.follow_symlinks,
-            exclude_patterns=args.exclude_patterns,
-            git=args.git,
-            unpack=not args.pack,
-            remove_sources=(args.contents is not None),
-            info=info,
-            worksheet_uuid=worksheet_uuid,
-            add_to_worksheet=True,
-        )
-        print >>self.stdout, uuid
+        # Create bundle
+        new_bundle = client.create('bundles', bundle_info, params={
+            'worksheet': worksheet_uuid
+        })
+
+        if args.contents is not None:
```

Comment:
Document more explicitly the ways people can upload by numbering them
1) Contents
2) URL
3) Local paths (usually one)